### PR TITLE
Create drawing tests for Core Text

### PR DESCRIPTION
**Pull request by aballway**
Cherry picks @DHowett-MSFT's  framework for image comparison tests and adds several tests for supported Core Text draw methods and options.  Fixes small issues found when comparing against reference platform results and when using custom fonts, but now compares against images generated by our platform for pixel comparison.  All differences between our platform and the reference platform are documented in #1502.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/1503)

### CODE REVIEW

File: tests/UnitTests/CoreGraphics.drawing/EntryPoint.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#include <windows.h>
+#include <TestFramework.h>
+#include <memory>
+
+#include <CoreFoundation/CoreFoundation.h>
+#include <UIKit/UIKit.h>
```
Comment (DHowett-MSFT) [inline tests/UnitTests/CoreGraphics.drawing/EntryPoint.cpp]:
```objc
#ifdef WINOBJC
// NOTE or whatever
#import <UIKit/UIKit.h>
static void anchorUIKit() { [UIView class]; }
#endif
```

**Comment by aballway** [comment]
Updated tests with custom font, squashed into two commits, and rebased on top of develop

**Comment by DHowett-MSFT** [comment]
When you merge this, please make sure to use the [Merge] Option. This will make conflict resolution in the CGD2D branch easier, as the commits have been cherry-picked.

**Comment by DHowett-MSFT** [comment]
The Rebase option may work as well.

**Comment by aballway** [comment]
Updated to disable tests on build machines, which seem to have minor differences in text drawing.  Not sure if this is worth investigating.  Tests can be run by enabling disabled tests

File: build/Tests/UnitTests/CoreGraphics.Drawing/CoreGraphics.Drawing.UnitTests.vcxproj [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    </ProjectReference>
+    <ProjectReference Include="..\..\..\CoreText\dll\CoreText.vcxproj">
+      <Project>{36deec5d-f77b-4c94-a63c-86fb716833de}</Project>
+    </ProjectReference>
+    <ProjectReference Include="..\..\..\AutoLayout\dll\AutoLayout.vcxproj">
+      <Project>{D036FDB1-F82C-40C7-B1B4-65F499EAE116}</Project>
+    </ProjectReference>
+    <ProjectReference Include="..\..\..\WinObjCRT\dll\WinObjCRT.vcxproj">
+      <Project>{585b4870-0d6b-43a6-8e7e-ad08f7f507b6}</Project>
+    </ProjectReference>
```
Comment (ms-jihua) [inline build/Tests/UnitTests/CoreGraphics.Drawing/CoreGraphics.Drawing.UnitTests.vcxproj]: haven't gotten to the rest of the review yet but right now i'm skeptical that you need this?

File: Frameworks/CoreGraphics/DWriteWrapper.mm [hunk 3]
```diff
@@ -355,14 +355,25 @@ HRESULT _DWriteCreateTextFormat(const wchar_t* fontFamilyName,
...
-                                           fontSize,
-                                           __GetUserDefaultLocaleName().data(),
-                                           outTextFormat); //
+    std::lock_guard<std::mutex> guard(_userCreatedFontCollectionMutex);
+    if (_userCreatedFontCollection) {
+        BOOL userCreatedFont = FALSE;
+        uint32_t unusedIndex;
+        RETURN_IF_FAILED(_userCreatedFontCollection->FindFamilyName(fontFamilyName, &unusedIndex, &userCreatedFont));
+        if (userCreatedFont) {
+            return dwriteFactory->CreateTextFormat(fontFamilyName,
```
Comment (ms-jihua) [inline Frameworks/CoreGraphics/DWriteWrapper.mm]: ah geez i need to think about this in the context of my CR, but, is this actually needed? shouldn't usercreatedfontcollection have been registered to the factory?

Comment (aballway) [inline Frameworks/CoreGraphics/DWriteWrapper.mm]: It's registered but this only uses the given font collection, or system font collection if nullptr

File: build/Tests/UnitTests/CoreGraphics.Drawing/CoreGraphics.Drawing.UnitTests.vcxproj (hunk 2 above)
Comment (ms-jihua) [inline build/Tests/UnitTests/CoreGraphics.Drawing/CoreGraphics.Drawing.UnitTests.vcxproj]: This reference

---
In reply to: [91159582](https://github.com/Microsoft/WinObjC/pull/1503#discussion_r91159582) [](ancestors = 91159582)

File: Frameworks/CoreGraphics/DWriteWrapper.mm (hunk 3 above)
Comment (ms-jihua) [inline Frameworks/CoreGraphics/DWriteWrapper.mm]: Grrr

---
In reply to: [91159166](https://github.com/Microsoft/WinObjC/pull/1503#discussion_r91159166) [](ancestors = 91159166)

File: tests/frameworks/OSXShims/include/windows.h [hunk 4]
```diff
@@ -25,6 +25,10 @@
 #define _MAX_PATH PATH_MAX
 #endif
 
+#ifndef MAX_PATH
+#define MAX_PATH PATH_MAX
```
Comment (ms-jihua) [inline tests/frameworks/OSXShims/include/windows.h]: er, any reason why we can't just standardize on the underscore version?

File: tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+//******************************************************************************
+
+#include "DrawingTest.h"
+
+struct CGNamedBlendMode {
+    const char* name;
+    CGBlendMode blendMode;
+};
+
+CGNamedBlendMode porterDuffBlendModes[] =
```
Comment (ms-jihua) [inline tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp]: static const, c_
same for your other constants below

File: tests/unittests/CoreGraphics.Drawing/CGContextDrawing_FillTests.cpp [hunk 6]
```diff
@@ -0,0 +1,105 @@
+//******************************************************************************
```
Comment (ms-jihua) [inline tests/unittests/CoreGraphics.Drawing/CGContextDrawing_FillTests.cpp]: This file has a _lot_ of duplicated code. Can you refactor it into reusable functions or parameterized tests?

File: tests/unittests/CoreGraphics.drawing/ImageHelpers.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    return url ? CFURLWriteDataAndPropertiesToResource(url.get(), data, nullptr, nullptr) : FALSE;
+}
+
+CFDataRef _CFDataCreateFromCGImage(CGImageRef image) {
+    return CGDataProviderCopyData(CGImageGetDataProvider(image));
+}
+
+CFStringRef _CFStringCreateAbsolutePath(CFStringRef relativePath) {
+    std::unique_ptr<char[]> owningFilenamePtr;
```
Comment (ms-jihua) [inline tests/unittests/CoreGraphics.drawing/ImageHelpers.cpp]: 85-101 here is repeated from above

**Comment by ms-jihua** [comment]
nit: CoreGraphics.Drawing > CoreGraphics.drawing, probably

**Comment by ms-jihua** [comment]
:clock1:

File: tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp (hunk 5 above)
Comment (aballway) [inline tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp]: Will leave the CG* tests for @DHowett-MSFT to clean up :D

File: tests/frameworks/OSXShims/include/windows.h (hunk 4 above)
Comment (aballway) [inline tests/frameworks/OSXShims/include/windows.h]: I believe @DHowett-MSFT said underscoreless is standard?

Comment (DHowett-MSFT) [inline tests/frameworks/OSXShims/include/windows.h]: Yeah: `MAX_PATH` lives in `minwindef` standalone and `_MAX_PATH` lives in stdlib with proximity to the standard library path functions.

File: tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp (hunk 5 above)
Comment (DHowett-MSFT) [inline tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp]: `O7`

File: tests/unittests/CoreGraphics.Drawing/CGContextDrawing_FillTests.cpp (hunk 6 above)
Comment (DHowett-MSFT) [inline tests/unittests/CoreGraphics.Drawing/CGContextDrawing_FillTests.cpp]: I'll take this on as I wrote this code. I'm hoping to reduce merge conflicts from develop in the meantime.

**Comment by DHowett-MSFT** [comment]
:shipit:

File: build/Tests/UnitTests/CoreGraphics.Drawing/CoreGraphics.Drawing.UnitTests.vcxproj (hunk 2 above)
Comment (aballway) [inline build/Tests/UnitTests/CoreGraphics.Drawing/CoreGraphics.Drawing.UnitTests.vcxproj]: Apparently required for SmartTypes #ByDesign

---
In reply to: [91159703](https://github.com/Microsoft/WinObjC/pull/1503#discussion_r91159703) [](ancestors = 91159703,91159582)

**Comment by aballway** [comment]
Updated with changes to CTDrawingTests

**Comment by DHowett-MSFT** [comment]
:shipit:

**Pull request merged**